### PR TITLE
feat: add GCE SD support in ScrapeConfig CRD

### PR DESCRIPTION
**Pull request by slashpai**
Fixes https://github.com/prometheus-operator/prometheus-operator/issues/6079

Related-to #5509

A few things to address:

As per https://prometheus.io/docs/prometheus/latest/configuration/configuration/#gce_sd_config , how credential is discovered is

```
Credentials are discovered by the Google Cloud SDK default client by looking in the following places, preferring the first location found:

a JSON file specified by the GOOGLE_APPLICATION_CREDENTIALS environment variable
a JSON file in the well-known path $HOME/.config/gcloud/application_default_credentials.json
fetched from the GCE metadata server
If Prometheus is running within GCE, the service account associated with the instance it is running on should have at least read-only permissions to the compute resources. If running outside of GCE make sure to create an appropriate service account and place the credential file in one of the expected locations.
```

**To clarify:**

Option 1: Should we let users update the Prometheus CR to include the secret and patch the container to set the ENV?
Option 2: Should we do Option 1 through operator?

Looking at how `FileSD` is implemented user needs to set the `configMap` in Prometheus CR and we are not setting it

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Add GCE SD support in ScrapeConfig CRD

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
feat: add GCE SD support in ScrapeConfig CRD
```

### CODE REVIEW

**Comment by slashpai** [comment]
As discussed in office going with option 1

**Comment by slashpai** [comment]
@simonpasquier could you please take a look at this one too? I can rebase along with addressing review comments if any 🙂

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 1]
```diff
@@ -390,3 +393,37 @@ type EC2SDConfig struct {
 	// +optional
 	Filters []*EC2Filter `json:"filters"`
 }
+
+// GCESDConfig allow retrieving scrape targets from GCE instances.
+// The private IP address is used by default, but may be changed to the public IP address with relabeling.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#gce_sd_config
+// Note: Users should update the Prometheus CR's spec.secret to include the Secret
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:400]: (suggested text)

The GCE service discovery will load the Google Cloud credentials from the file specified by the  `GOOGLE_APPLICATION_CREDENTIALS` environment variable.
A pre-requisite for using `GCESDConfig` is that a Secret containing valid Google Cloud credentials is mounted into the Prometheus or PrometheusAgent pod via the `.spec.secrets` field and that the `GOOGLE_APPLICATION_CREDENTIALS` environment variable is set to `/etc/prometheus/secrets/<secret-name>/<credentials>`.

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -2702,6 +2702,60 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		})
 	}
 
+	// GCESDConfig
+	if len(sc.Spec.GCESDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.GCESDConfigs))
+		for i, config := range sc.Spec.GCESDConfigs {
+			if config.Project != "" {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2709]: Project can't be empty by the API definition.

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -2702,6 +2702,60 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+				configs[i] = []yaml.MapItem{
+					{
+						Key:   "project",
+						Value: config.Project,
+					},
+				}
+			}
+
+			if config.Zone != "" {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2718]: same for zone

**Comment by simonpasquier** [comment]
Looking at https://cloud.google.com/kubernetes-engine/docs/tutorials/authenticating-to-cloud-platform this seems to be the correct approach.

**Comment by slashpai** [comment]
@simonpasquier addressed comments, please review :)

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 4]
```diff
@@ -429,3 +432,43 @@ type AzureSDConfig struct {
 	// +optional
 	Port *int `json:"port"`
 }
+
+// GCESDConfig  allow retrieving scrape targets from GCP GCE instances
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:436]:
```suggestion
// GCESDConfig  configures scrape targets from GCP GCE instances.
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 5]
```diff
@@ -429,3 +432,43 @@ type AzureSDConfig struct {
...
+// the public IP address with relabeling.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#gce_sd_config
+//
+// The GCE service discovery will load the Google Cloud credentials
+// from the file specified by the GOOGLE_APPLICATION_CREDENTIALS environment variable.
+// See https://cloud.google.com/kubernetes-engine/docs/tutorials/authenticating-to-cloud-platform
+//
+// A pre-requisite for using GCESDConfig is that a Secret containing valid
+// Google Cloud credentials is mounted into the Prometheus or PrometheusAgent
+// pod via the .spec.secrets field and that the GOOGLE_APPLICATION_CREDENTIALS
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:447]:
```suggestion
// pod via the `.spec.secrets` field and that the GOOGLE_APPLICATION_CREDENTIALS
```

**Review by simonpasquier — COMMENTED** [review]
2 small final nits.

**Pull request merged**